tests.contrib.tornado.test_safety.TestAsyncConcurrency.test_concurrent_requests fails randomly

```
=================================== FAILURES ===================================
________________ TestAsyncConcurrency.test_concurrent_requests _________________

self = <tests.contrib.tornado.test_safety.TestAsyncConcurrency testMethod=test_concurrent_requests>

    @gen_test
    def test_concurrent_requests(self):
        # the application must handle concurrent calls
        def make_requests():
            # use a blocking HTTP client (we're in another thread)
            http_client = httpclient.HTTPClient()
            url = self.get_url('/nested/')
            response = http_client.fetch(url)
            assert 200 == response.code
            assert 'OK' == response.body.decode('utf-8')
            # freeing file descriptors
            http_client.close()
    
        # blocking call executed in different threads
        threads = [threading.Thread(target=make_requests) for _ in range(25)]
        for t in threads:
            t.daemon = True
            t.start()
    
        # wait for the execution; assuming this time as a timeout
        yield web.compat.sleep(0.5)
    
        # the trace is created
        traces = self.tracer.writer.pop_traces()
>       assert 25 == len(traces)
E       AssertionError: assert 25 == 14
E        +  where 14 = len([[<Span(id=6137145678679891056,trace_id=7032705444916804637,parent_id=None,name=tornado.request)>, <Span(id=8545696739...>, <Span(id=16292201187504429261,trace_id=9654555000514619975,parent_id=5310689741135979469,name=tornado.sleep)>], ...])

tests/contrib/tornado/test_safety.py:42: AssertionError
```

https://circleci.com/gh/DataDog/dd-trace-py/121252

--- a/tests/contrib/tornado/test_safety.py
+++ b/tests/contrib/tornado/test_safety.py
@@ -17,29 +17,35 @@ class TestAsyncConcurrency(TornadoTestCase):
     """
     @gen_test
     def test_concurrent_requests(self):
+        REQUESTS_NUMBER = 25
+        responses = []
+
         # the application must handle concurrent calls
         def make_requests():
             # use a blocking HTTP client (we're in another thread)
             http_client = httpclient.HTTPClient()
             url = self.get_url('/nested/')
             response = http_client.fetch(url)
+            responses.append(response)
             assert 200 == response.code
             assert 'OK' == response.body.decode('utf-8')
             # freeing file descriptors
             http_client.close()
 
         # blocking call executed in different threads
-        threads = [threading.Thread(target=make_requests) for _ in range(25)]
+        threads = [threading.Thread(target=make_requests) for _ in range(REQUESTS_NUMBER)]
         for t in threads:
-            t.daemon = True
             t.start()
 
-        # wait for the execution; assuming this time as a timeout
-        yield web.compat.sleep(0.5)
+        while len(responses) < REQUESTS_NUMBER:
+            yield web.compat.sleep(0.001)
+
+        for t in threads:
+            t.join()
 
         # the trace is created
         traces = self.tracer.writer.pop_traces()
-        assert 25 == len(traces)
+        assert REQUESTS_NUMBER == len(traces)
         assert 2 == len(traces[0])
 
 

--- a/tests/contrib/tornado/web/compat.py
+++ b/tests/contrib/tornado/web/compat.py
@@ -1,4 +1,5 @@
 from tornado.concurrent import Future
+import tornado.gen
 from tornado.ioloop import IOLoop
 
 
@@ -24,12 +25,16 @@ except ImportError:
             super(ThreadPoolExecutor, self).__init__()
 
 
-def sleep(duration):
-    """
-    Compatibility helper that return a Future() that can be yielded.
-    This is used because Tornado 4.0 doesn't have a ``gen.sleep()``
-    function, that we require to test the ``TracerStackContext``.
-    """
-    f = Future()
-    IOLoop.current().call_later(duration, lambda: f.set_result(None))
-    return f
+if hasattr(tornado.gen, 'sleep'):
+    sleep = tornado.gen.sleep
+else:
+    # Tornado <= 4.0
+    def sleep(duration):
+        """
+        Compatibility helper that return a Future() that can be yielded.
+        This is used because Tornado 4.0 doesn't have a ``gen.sleep()``
+        function, that we require to test the ``TracerStackContext``.
+        """
+        f = Future()
+        IOLoop.current().call_later(duration, lambda: f.set_result(None))
+        return f

